### PR TITLE
[image] Add responsive web images

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

We want to select image source that best fits the container

# How

https://user-images.githubusercontent.com/5597580/207411153-31eb0498-3aae-4364-ab8c-29cb325d3662.mov

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 1]
```diff
@@ -150,6 +150,80 @@ const useTransition = (
...
+};
+
+const useSourceSelection = (
+  sources?: ImageSource[],
+  sizeCalculation: 'initial' | 'live' = 'initial'
+) => {
+  const hasMoreThanOneSource = (sources?.length ?? 0) > 1;
+
+  // undefined - not calculated yet, don't fetch any images, null - no size available, pick arbitrary image, DOMRect - size available
+  const [size, setSize] = React.useState<null | undefined | DOMRect>(undefined);
```
Comment (aleqsio) [inline packages/expo-image/src/ExpoImage.web.tsx:185]: @tsapeta ok or ugly?

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 2]
```diff
@@ -150,6 +150,80 @@ const useTransition = (
...
+  size: DOMRect | null
+): ImageSource | null => {
+  return (
+    sources
+      // look for the smallest image that's still larger then a container
+      ?.map((source) => {
+        if (!size) return { source, penalty: 0, covers: false };
+        const { width, height } =
+          typeof source === 'object' ? source : { width: null, height: null };
+        if (width == null || height == null) return { source, penalty: 0, covers: false };
```
Comment (lukmccall) [inline packages/expo-image/src/ExpoImage.web.tsx:164]:
```suggestion
        if (width == null || height == null) {
          return { source, penalty: 0, covers: false };
        }
```
I'm not a big fun of the if statement without brackets, but is up to you ;)

File: packages/expo-image/src/Image.types.ts [hunk 3]
```diff
@@ -108,6 +108,12 @@ export type ImageProps = AccessibilityProps & {
    */
   cachePolicy?: ImageCachePolicy | null;
 
+  /**
+   * Determines whether to choose image source based on container size only on mount or on every resize.
+   * @platform web
+   */
+  webResponsivePolicy?: 'live' | 'initial';
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:115]: I would remove `web`, I don't know about any platform-specific prop that we prefix with `ios` or `android` 🤔 

```suggestion
  responsivePolicy?: 'live' | 'initial';
```

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 4]
```diff
@@ -150,6 +150,80 @@ const useTransition = (
...
+        return { source, penalty: (width - size.width) * (height - size.height), covers: true };
+      })
+      .sort((a, b) => a.penalty - b.penalty)
+      .sort((a, b) => Number(b.covers) - Number(a.covers))[0]?.source ?? null
+  );
+};
+
+const useSourceSelection = (
+  sources?: ImageSource[],
+  sizeCalculation: 'initial' | 'live' = 'initial'
```
Comment (tsapeta) [inline packages/expo-image/src/ExpoImage.web.tsx:180]:
```suggestion
  sizeCalculation: ImageProps['responsivePolicy'] = 'initial'
```

File: packages/expo-image/src/ExpoImage.web.tsx (hunk 2 above)
Comment (aleqsio) [reply packages/expo-image/src/ExpoImage.web.tsx:164]: Yeah, Tomek already mentioned that 😢

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 5]
```diff
@@ -150,6 +150,80 @@ const useTransition = (
...
+        if (width < size.width || height < size.height)
+          return {
+            source,
+            penalty: Math.max(size.width - width, size.height - height),
+            covers: false,
+          };
+        return { source, penalty: (width - size.width) * (height - size.height), covers: true };
+      })
+      .sort((a, b) => a.penalty - b.penalty)
+      .sort((a, b) => Number(b.covers) - Number(a.covers))[0]?.source ?? null
```
Comment (tsapeta) [inline packages/expo-image/src/ExpoImage.web.tsx:174]: `.sort` in JS is in-place, so `findBestSourceForSize` has a side effect but its name doesn't suggest that 😅 I think it's better to copy the array and keep the input unsorted.

File: packages/expo-image/src/Image.types.ts [hunk 6]
```diff
@@ -108,6 +108,14 @@ export type ImageProps = AccessibilityProps & {
    */
   cachePolicy?: ImageCachePolicy | null;
 
+  /**
+   * Determines whether to choose image source based on container size only on mount or on every resize.
+   * Use `live` to improve performance.
+   * @default "initial"
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:114]: You've changed it to `live` in `useSourceSelection` 😅

Comment (aleqsio) [reply packages/expo-image/src/Image.types.ts:114]: See, this is why I can't code and talk 🤦

**Review by lukmccall — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

**Comment by erickreutz** [comment]
I'm wondering how we go about using Expo Image to load retina assets on the web? The documentation is pretty sparse and the traditional way of React Native loading the retina asset based on the pixel ratio, which happens automatically on native, doesn't seem to exist on the web? What is the recommended way of accomplishing this with Expo Image on the web?

**Pull request merged**